_posts/2016-04-26-distributed-game-of-life-in-elixir.md

```diff
@@ -0,0 +1,1065 @@
+---
+
+I wrote my first game in Elixir. It's a popular thing - Conway's Game of Life - but it gets quite interesting when you solve it in a functional language, especially when you can see how [actor model](https://en.wikipedia.org/wiki/Actor_model) works and how actors are distributed across servers in network.
+
+<img src="/images/blog/posts/distributed-game-of-life-in-elixir/game_of_life_logo.png" style="float:right;margin-left:20px;" />
+
+__What I'm going to show you is:__
+
+* How to write rules for game of life with tests in Elixir
+* Parallel tasks across lightweight processes (actors) so we utilize all CPU cores
+* How to distribute work across nodes so the game can be executed by many servers in the cluster
+* How to use GenServer behaviour, TaskSupervisor and Agents in Elixir
+
+This project and the full [source code can be found here](https://github.com/BeyondScheme/elixir-game_of_life).
+
+# Demo
+
+Let's start with watching quick demo how the game works.
+
+<script type="text/javascript" src="https://asciinema.org/a/44233.js" id="asciicast-44233" async data-preload="true"></script>
+
+As you see we started node1 where the game is running and board is printed on the screen. We started also the second node and connected it to the first one. From the second node, we added new cells to the board. Both nodes are responsible for processing the game but only the first node is a master with info about the current state of the game. We can connect more nodes to the cluster so game processing can happen on all of the nodes. You are going to learn in this article how to make it happen.
+
+# Game of Life rules
+
+If you haven't heard about [game of life problem](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life) then here is basic concept. If you already know it, just jump to [next header](#create-new-application-in-elixir).
+
+The universe of the Game of Life is an infinite two-dimensional orthogonal grid of square cells, each of which is in one of two possible states, alive or dead. Every cell interacts with its eight neighbours, which are the cells that are horizontally, vertically, or diagonally adjacent. At each step in time, the following transitions occur:
+
+* Any live cell with fewer than two live neighbours dies, as if caused by under-population.
+* Any live cell with two or three live neighbours lives on to the next generation.
+* Any live cell with more than three live neighbours dies, as if by over-population.
+* Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+
+The initial pattern constitutes the seed of the system. The first generation is created by applying the above rules simultaneously to every cell in the seed—births and deaths occur simultaneously, and the discrete moment at which this happens is sometimes called a tick (in other words, each generation is a pure function of the preceding one). The rules continue to be applied repeatedly to create further generations.
```

Comment:
BTW this text is from Wikipedia. I was wondering about adding it in the quote but it looks bad when you have a wall of text as the quote so instead, I added the link to the source on wiki in the fist sentence for this chapter.